### DISCUSSION TITLE
stdin

### DISCUSSION THREAD

**Discussion by kkoomen** [pyodide/pyodide#4645] [General]
Hi,

I've tried to get stdin working but without any success, unfortunately. I've tried the following:

```py
import sys

def get_input(prompt):
    sys.stdout.write(prompt)
    sys.stdout.flush()
    return sys.stdin.readline().rstrip()

x = get_input('Type something: ')
print(x)
```

but this still requests for input and then prints it, while I expect to first print what's been written to the stdout _and then_ request for input. Is this possible with the latest version of Pyodide?

**Comment by ryanking13** [comment]
Thanks for the report. I think you can additionally do `os.fsync(sys.stdout.fileno())` to ensure the data is written ([ref](https://docs.python.org/3/library/os.html#os.fsync)) to the file (in this case, stdout).

**Reply by kkoomen** [reply]
If I change `sys.stdout.flush()` with `os.fsync(sys.stdout.fileno())` inside the code I provided above, I get the following error:

```
Traceback (most recent call last):
  File "/lib/python311.zip/_pyodide/_base.py", line 499, in eval_code
    .run(globals, locals)
     ^^^^^^^^^^^^^^^^^^^^
  File "/lib/python311.zip/_pyodide/_base.py", line 340, in run
    coroutine = eval(self.code, globals, locals)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<exec>", line 8, in <module>
  File "<exec>", line 5, in get_input
io.UnsupportedOperation: fileno
```

**Reply by ryanking13** [reply]
Are you running the code in the Pyodide console? If so, please try running it somewhere, not in the Pyodide console. Pyodide console uses a [very simplified WriteStream](https://github.com/pyodide/pyodide/blob/c776168c5351a20169d9fbee6f722f64ba69bd41/src/py/pyodide/console.py#L42-L58), which does not implement all methods.

**Reply by kkoomen** [reply]
@ryanking13 I'm simply executing the code through `pyodide.runPython()`, is stdin supported through this?

**Reply by ryanking13** [reply]
Yes, stdout is supported and it should work (I guess you mean stdout not stdin?). What do you get if you do `print(sys.stdout)`? If should be

```
<_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
```

and

```
import os, sys
os.fsync(sys.stdout.fileno())
```

should work

**Reply by kkoomen** [reply]
Ah, so what I do inside our project is: I have an editor and the user can type in some Python code and run it. When it will be executed, I always run `sys.stdout = io.StringIO()` first, then execute the code and then run `sys.stdout.getvalue()`, but I guess that's not an option? How would I use `input()` without the use of `sys.stdout = io.StringIO()`? Any recommendations?

**Reply by ryanking13** [reply]
> How would I use input() without the use of sys.stdout = io.StringIO()

Well, `input()` is not related to stdout but stdin, so I am unsure if I understood your question correctly.

Anyway, we provide some way to redirect streams: https://pyodide.org/en/stable/usage/streams.html, so you can redirect stdin and stdout. I guess this document will help?

**Reply by kkoomen** [reply]
@ryanking13 Okay so let me just clarify what I actually want in pyodide.

I want to run this script exactly the same in pyodide as if I were to run this in a python3 REPL or script:

```py
print('Welcome')
num1 = input('Enter a number: ')
print('Thank you!')
num2 = input('Enter another number: ')
print(f'Your two numbers are: {num1} and {num2}')
```

but currently our version outputs this:


https://github.com/pyodide/pyodide/assets/10693490/d9a16d35-76ec-45b5-ac2a-53dcc216efe8

FYI: we're using pyodide inside a worker. Here's the relevant code: https://github.com/uvapl/examide/blob/feature/stdin-py/static/js/workers/py.worker.js#L20-L69

Any idea on how to make this work properly? Ideally, the output should be this:

```
Welcome
Enter a number: 123
Thank you!
Enter another number: 456
Your two numbers are: 123 and 456
```

**Reply by ryanking13** [reply]
The `input()` calls fflush to flush the batched output, but it seems like that is not enough, and you need to call `fsync` to actually apply the output to disk (in this case, stdout). TBH, I don't really understand the difference between the two, but there seems a difference (https://github.com/emscripten-core/emscripten/issues/14457#issuecomment-862486220).

So I think we have two options here:

1. Use "unbatched" stdout.

By default, when you pass `stdout` parameter when loading pyodide. We use "batch" mode. That is, the data is only flushed when it gets "\n".

We have a "raw" mode that does not use batching, which I think you can have a look.

```js
// see: https://pyodide.org/en/stable/usage/api/js-api.html#pyodide.setStdout
pyodide.setStdout({ raw: handler })
```

2. Find a way to call `fsync` inside a `input`.

You can override the builtin `input` function to call `os.fsync(sys.stdout.fileno())`.

**Reply by kkoomen** [reply]
Oh! that indeed fixed it! Awesome, thanks a lot for your time, I appreciate it!


https://github.com/pyodide/pyodide/assets/10693490/874d6add-3bad-47b7-8f58-d77cb39b564b

**Reply by kkoomen** [reply]
@ryanking13 We just ran into a new issue using the "unbatched" mode (i.e. raw stdout handler). When running `print('hello', end="")` the raw stdout callback is not called and also the `pyodide.runPython(code)` does not return anything (i.e. returns `undefined`). Any idea how this can be handled properly?

**Reply by hoodmane** [reply]
Python does line buffering on stdout by default. You'll need to call `sys.stdout.flush()` or call `loadPyodide` with `{args: ["-u"]}` to disable buffering.

**Reply by kkoomen** [reply]
@hoodmane Thanks for responding quickly. The `-u` flag indeed solves the problem!
